src/lib/core/TLVWriter.cpp

```diff
@@ -97,6 +97,22 @@ CHIP_ERROR TLVWriter::Finalize()
     return err;
 }
 
+CHIP_ERROR TLVWriter::ReserveBuffer(uint32_t aBufferSize)
+{
+    VerifyOrReturnError(mRemainingLen >= aBufferSize, CHIP_ERROR_NO_MEMORY);
+
+    uint32_t spaceLeftInCurrentBuffer        = mReservedSize + mRemainingLen;
+    bool canPossiblyAllocateAdditionalMemory = spaceLeftInCurrentBuffer < mMaxLen;
+
+    if (mBackingStore && canPossiblyAllocateAdditionalMemory)
+    {
+        VerifyOrReturnError(mBackingStore->IsSafeToReserve(), CHIP_ERROR_INCORRECT_STATE);
+    }
+    mReservedSize += aBufferSize;
+    mRemainingLen -= aBufferSize;
```

Comment:
The logic of reserving space in the current buffer just seems wrong when there is a backing store that might (or might not) be able to allocate more memory later. Shouldn't ReserveBuffer effectively just be delegated to the backing store, and then it's up to the backing store to decide if it that amount of memory can be reserved?

`ReserveBuffer` is also rather under-documented... Is it intended to guarantee that a `CHIP_ERROR_NO_MEMORY` will not be encountered later during writing, after un-reserving the space? (The UnreserveBuffer API is also ambiguous as to whether the caller wants to give up the reserved space because it's not needed after all, or wants to actually start using the reserved space.)

Reply:
The interactions of ReserveBuffer with a backing store just seems broken in a number of ways currently... WriteData also ends up under-filling the current buffer because the space that was reserved is left unused.

Reply:
I think the correct approach might be for reservation to temporarily decrease `mMaxLen`, i.e. reserve logical space that can be written later, not physical space to leave empty at the end of the current buffer. This means the purpose of reservation is not to prevent NO_MEMORY errors later, but to keep track of a final target size (e.g. of a packet) we don't want to exceed, independently of the layout of the underlying buffers.

Reply:
If we were writing this API from the start and didn't have to account for how it is used today, I would 100% agree with you. Problem is now mentioned in the description of the PR that for legacy usage of `TLVWriter::ReserveBuffer`

My objective for this PR is to disable the overrun footgun while not breaking existing usage of `TLVWriter::ReserveBuffer`

Reply:
This looks OK to me as an immediate fix.